cluster/mesos/docker/util.sh

```diff
@@ -150,6 +149,37 @@ function cluster::mesos::docker::run_in_docker_cagen {
   return "${exit_status}"
 }
 
+# Run kube-keygen.sh inside docker.
+function cluster::mesos::docker::run_in_docker_keygen {
```

Comment:
According to https://github.com/docker/docker/issues/9351#issuecomment-158262910 the docker PID 1 only responds to SIGSTOP & SIGKILL. It is designed to be used with `docker stop` and `docker kill`. So even if you use `exec docker run -it` PID 1 (by default) ignores SIGTERM and SIGINT.  That's why we're trapping SIGTERM & SIGINT and translating them into `docker stop` which will try to gracefully stop PID 1 and then kill it if it's still up after a timeout.

Apparently, it's relatively easy to make bash entrypoint scripts listen for SIGTERM and SIGINT. Just add a trap. Even if the trap doesn't do anything, bash seems to pass the signal to the child process that's running:

```
$ docker run --rm -it ubuntu:14.04.3 bash -ceu "trap - INT; sleep 30"
^C$
```

This appears to work even running the docker cli in a docker container:

```
$ docker run --rm -it -v "/var/run/docker.sock:/var/run/docker.sock" mesosphere/kubernetes-mesos-test -ceux "trap 'echo SIGINT-outer' INT; docker run --rm -it ubuntu:14.04.3 bash -ceux \"trap 'echo SIGINT-inner' INT; sleep 30\""
+ trap 'echo SIGINT-outer' INT
+ docker run --rm -it ubuntu:14.04.3 bash -ceux 'trap '\''echo SIGINT-inner'\'' INT; sleep 30'
+ trap 'echo SIGINT-inner' INT
+ sleep 30
^C++ echo SIGINT-inner
SIGINT-inner
$
```

Interestingly, the "SIGINT-outer" trap is not required. As long as the "SIGINT-inner" trap exists, signal handling is enabled.

So it can be interrupted even before the inner `docker run` executes:

```
$ docker run --rm -it -v "/var/run/docker.sock:/var/run/docker.sock" mesosphere/kubernetes-mesos-test -ceux "sleep 30; docker run --rm -it ubuntu:14.04.3 bash -ceux \"trap 'echo SIGINT-inner' INT; sleep 30\""
+ sleep 30
^C$
```

Based on this experimentation it doesn't look like `exec` is required in combination with `docker run -it`, as long as there's a SIGINT & SIGTERM trap somewhere in the entrypoint. Weird!